### ISSUE TITLE
[Bug]: babel/preset-env - document says esmodules is intersected, build warning says its ignored.

### ISSUE THREAD

**Issue by epreston** [babel/babel#15371]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

babel-loader (webpack)

### Input code

```js
console.log('hello world!');
```

### Configuration file name

.babelrc.json

### Configuration

```json
{
  "presets": [
    [
      "@babel/preset-env",
      {
        "bugfixes": true,
        "targets": {
          "browsers": "Edge >= 108, Firefox >= 108, FirefoxAndroid  >= 107, Chrome >= 108, ChromeAndroid >= 108, Safari >= 16.1, iOS >= 16.1, not dead",
          "esmodules": true
        },
        "useBuiltIns": "usage",
        "corejs": "3.27.2"
      }
    ]
  ]
}
```

### Current and expected behavior

Documentation for [options "targets"](https://babeljs.io/docs/en/options#targets) says:

> Please note: when specifying both browsers and the esmodules target, they will be intersected.

However, when building the following message is displayed.  

> @babel/preset-env: esmodules and browsers targets have been specified together.
>`browsers` target, `Edge >= 108, Firefox >= 108, FirefoxAndroid  >= 107, Chrome >= 108, ChromeAndroid >= 108, Safari >= 16.1, iOS >= 16.1, not dead` will be ignored.

Which is correct ?  Is it intersected or ignored ?

### Environment

Babel Versions: 
  - "@babel/core": "^7.20.12",
  - "@babel/preset-env": "^7.20.2",
  - "babel-loader": "^9.1.2",

Node Versions:
  - npm 8.19.3
  - node 18.13.0

OS:
  - WIndows 10
  - OSX (latest)

### Possible solution

Depending on which is correct, update documentation or warning message in `babel/preset-env`

### Additional context

I would like rely on the intersection behaviour in the future, if possible.

**Comment by epreston** [comment]
Thank you babel-bot.  We need to update the documentation or tool.

**Comment by nicolo-ribaudo** [comment]
It should definitely be intersected, I'll check what the code does.

**Comment by nicolo-ribaudo** [comment]
If you specify your targets at the top level of your config (so in the generic Babel options — https://babeljs.io/docs/en/options#targetsesmodules), `esmodules` will be intersected with `browsers`:
```json
{
  "targets": {
    "browsers": "Edge >= 108, Firefox >= 108, FirefoxAndroid  >= 107, Chrome >= 108, ChromeAndroid >= 108, Safari >= 16.1, iOS >= 16.1, not dead",
     "esmodules": true
  },
  "presets": [
    [
      "@babel/preset-env",
      {
        "bugfixes": true,
        "useBuiltIns": "usage",
        "corejs": "3.27.2"
      }
    ]
  ]
}
```

However, for backward compatibility we couldn't change the pre-existing behavior of `@babel/preset-env`'s `targets` option, so `esmodules` there completely overwrites `browsers`. We need to document this.

I recommend using the top-level `targets` option, instead of `@babel/preset-env`'s `targets`, because:
- it has the better intersection behavior
- can also be read by other plugins that can adjust their behavior based on your targets, while if you use it just in `@babel/preset-env` other plugins will always assume the oldest possible targets.

**Comment by liuxingbaoyu** [comment]
> We need to document this.

We can also modify the message in code to link to the documentation.

**Comment by epreston** [comment]
I see, that makes sense.  You are right, using the top-level targets option is the best way to go.  

Should I leave this open to capture the "documentation" task ?

**Comment by nicolo-ribaudo** [comment]
> Should I leave this open to capture the "documentation" task ?

Yes please! (or, if you want to open a PR to github.com/babel/website, that would be even better :eyes:)

**Comment by epreston** [comment]
Good idea, I will try my hand at updating the documentation.  I'll reference the info you provided here and close issue once the PR is submitted.

**Comment by epreston** [comment]
PR Submitted.  Thank you for clarifying this for me.

**Comment by epreston** [comment]
> Yes please! (or, if you want to open a PR to github.com/babel/website, that would be even better 👀)

I did that

**Comment by nicolo-ribaudo** [comment]
Sorry, the only problem was that I somehow forgot about your PR 😅

**Comment by epreston** [comment]
Maybe it's for the best.  @JLHwung was able to help us refine it with the extra time.

**Comment by lazka** [comment]
The situation is more confusing now.

`esmodules` now has a "intersect" value, and it works as documented (debug shows that esmodules and browsers are intersected):

```
                    presets: [
                        [
                            '@babel/preset-env',
                            {
                                loose: false,
                                shippedProposals: true,
                                bugfixes: true,
                                modules: false,
                                debug: true,
                                targets: {
                                    esmodules: 'intersect',
                                    browsers: 'defaults and not dead, last 5 years',
                                },
                            },
                        ],
```

but you still get a warning stating that it's ignored, while that is not true:

```
@babel/preset-env: esmodules and browsers targets have been specified together.
`browsers` target, `defaults and not dead, last 5 years` will be ignored.
```

Ideally the warning would not be shown there.

**Comment by JLHwung** [comment]
@lazka Yes, you are right. We overlooked the fact that `esmodules` can be `intersect` as well:

https://github.com/babel/babel/blob/5051613c021b6604929278d6811fa671f49dfbdb/packages/babel-preset-env/src/index.ts#L277-L282

Do you want to open a PR?
